### PR TITLE
[operator] fix libentry to support triton 2.3

### PR DESCRIPTION
**Pull request by StrongSpoon**
1. modified libentry arguments collection, cache key generation and constexpr collection
2. specify dns arguments for operators
3. works for triton 2.2 and 2.3

### CODE REVIEW

File: src/flag_gems/ops/dropout.py [hunk 1]
```diff
@@ -82,7 +82,7 @@ def dropout_forward_kernel(
         "N",
     ],
 )
-@triton.jit
+@triton.jit(do_not_specialize=["p", "philox_seed", "philox_offset"])
```
Comment (tongxin) [inline src/flag_gems/ops/dropout.py:85]: Does Triton specialize on floats?

File: src/flag_gems/utils/libentry.py [hunk 2]
```diff
@@ -14,47 +16,110 @@ def __init__(
         while not isinstance(fn, triton.runtime.JITFunction):
             fn = fn.fn
         self.jit_function: triton.runtime.JITFunction = fn
-        self.kernel_arg_indices = []
+        self.spec_indices = []
+        self.dns_indices = []
```
Comment (tongxin) [inline src/flag_gems/utils/libentry.py]: let's be more precise? do_not_specialize_indices

File: src/flag_gems/utils/libentry.py [hunk 3]
```diff
@@ -14,47 +16,110 @@ def __init__(
...
         for p in self.jit_function.params:
             if not p.is_constexpr:
-                self.kernel_arg_indices.append(p.num)
+                if p.do_not_specialize:
+                    self.dns_indices.append(p.num)
+                else:
+                    self.spec_indices.append(p.num)
```
Comment (tongxin) [inline src/flag_gems/utils/libentry.py]: Isn't a bit of list comprehension is more favorable than loops?

File: src/flag_gems/utils/libentry.py [hunk 4]
```diff
@@ -14,47 +16,110 @@ def __init__(
...
-        for arg in args:
+    def key(self, spec_args, dns_args, const_args):
+        entry_key = []
+        for arg in spec_args:
+            if hasattr(arg, "data_ptr"):
+                entry_key.append(str(arg.dtype))
+                entry_key.append(arg.data_ptr() % self.divisibility == 0)
+            else:
+                entry_key.append(type(arg))
+                entry_key.append(arg)
```
Comment (tongxin) [inline src/flag_gems/utils/libentry.py]: List comprehension is an immutable fashion for coding loops. It's normally more efficient than the mutable counterpart.

File: src/flag_gems/ops/dropout.py (hunk 1 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/dropout.py:85]: floats not marked as dns are specialized as False

File: src/flag_gems/utils/libentry.py (hunk 2 above)
Comment (StrongSpoon) [reply src/flag_gems/utils/libentry.py]: done

File: src/flag_gems/utils/libentry.py (hunk 3 above)
Comment (StrongSpoon) [reply src/flag_gems/utils/libentry.py]: done

File: src/flag_gems/utils/libentry.py (hunk 4 above)
Comment (StrongSpoon) [reply src/flag_gems/utils/libentry.py]: done

**Comment by iclementine** [comment]
LGTM

The triton jit runtime is too complicated. 
- It offers several ways to pass parameter(via explicitly passing parameter, using default values, via decorator cascading(autotunner & heuristics).
- It offers several ways to specify how arguments are handled in the jit decorator (do not specialize, ...) or in the function (tl.constexpr). 
- It handles different types of actual arguments differently. (specialization of different types of arguments)
- It has some limitations of argument types, which is not well-documented.
- The parser handles values differntly depending on whether a value is a constexpr or tensor, while these processes are not well-documented.(equal to 1 arguments are handled as if they were constexpr)

After all, for quick understanding of the rationale, a cheatsheet is needed.
图片

**Review by iclementine — APPROVED** [review]

**Pull request merged**